### DISCUSSION TITLE
np.datetime64('NAT') fill value

### DISCUSSION THREAD

**Discussion by jvail** [pydata/xarray#5200] [Q&A]
I do not understand why this (xarray 0.17.0)

```
da = xr.DataArray([np.datetime64('NAT')])
da.to_netcdf('da')
```
is OK and the following gives an error

```
da = xr.DataArray([np.datetime64('NAT')], attrs={'_FillValue': np.datetime64('NAT')})
da.to_netcdf('da')
```

`TypeError: Invalid value for attr '_FillValue': numpy.datetime64('NaT') must be a number, a string, an ndarray or a list/tuple of numbers/strings for serialization to netCDF files`

What is a "proper" fill value for datedime64? Thank you, Jan

**Comment by max-sixty** [comment]
IIUC, `_FillValue` is just a string there — it raises an error because the values of `attrs` must be one of those types.

Is there a specific reason `_FillValue` needs to be in the attrs? (I'm not a big netcdf user so there may be)

**Reply by max-sixty** [reply]
Is the intention for this to go onto the encoding attribute? http://xarray.pydata.org/en/stable/io.html#scaling-and-type-conversions

**Reply by jvail** [reply]
Sorry @max-sixty, a bit more context: I am using the library xarray-simlab (https://github.com/xarray-contrib/xarray-simlab) and it stores the `_FillValue` as an attribute with in turn is used by netcdf. When I change it to a string I get this error from netcdf:

```
   1031         if '_FillValue' in self._attributes:
-> 1032             fill_value = np.array(self._attributes['_FillValue'],
   1033                                   dtype=self.data.dtype).tobytes()
   1034             if len(fill_value) == self.itemsize():

ValueError: could not convert string to float: b'NAT'
```

So, I wonder what the appropriate, working fill value for datetime64 in xarray might be. 'NAT' seems a good choice - just does not work.

**Reply by max-sixty** [reply]
Could you link to docs / code on where it stores `_FillValue` as at `attrs`?

**Reply by dcherian** [reply]
> ValueError: could not convert string to float: b'NAT'

Times are stored as numbers representing "hours since xxxx" (for e.g.). So you cannot assign a string for `_FillValue`. Why are you changing it?

**Reply by jvail** [reply]
> > ValueError: could not convert string to float: b'NAT'
> 
> Times are stored as numbers representing "hours since xxxx" (for e.g.). So you cannot assign a string for `_FillValue`. Why are you changing it?

I just tested a str because if I use a datetime64 xarray complains and if I use a str xarray is happy but netcdf fails. The confusing part is: If I can use a value like datetime64('NAT') inside an array why can I not use it as a fill value?

**Reply by dcherian** [reply]
> datetime64('NAT') inside an array why can I not use it as a fill value?

Because netCDF does not accept it. it has no understanding of datetime types so we "encode" it to a number. The in-memory `NaT` will be represented by `_FillValue` on disk (and it must be a number).

Why are you trying to set `_FillValue`? Is the automatic choice not good enough?

**Reply by jvail** [reply]
> Why are you trying to set `_FillValue`? Is the automatic choice not good enough?

It is actually set by xarray-simlab. I just tried to make a little test case to figure out why saving my xarray dataset (the ouput of xarray-simlab) fails.

> Because netCDF does not accept it. it has no understanding of datetime types so we "encode" it to a number. The in-memory `NaT` will be represented by `_FillValue` on disk (and it must be a number).

Then it might be an xarray issue - this is why I ask the question here :). Why does xarray accept 'NAT' as a value but not as a fill value when I want to save it as netcdf? I don't get it - sorry, maybe I am too slow. I would expect that anything that can be a value can be a fill value.
